src/workerd/api/streams/readable-source.c++

```diff
@@ -0,0 +1,852 @@
+#include "readable-source.h"
+
+#include "writable-sink.h"
+
+#include <workerd/api/util.h>
+#include <workerd/io/io-context.h>
+#include <workerd/jsg/jsg.h>
+#include <workerd/util/stream-utils.h>
+#include <workerd/util/string-buffer.h>
+#include <workerd/util/strong-bool.h>
+
+#include <kj/async-io.h>
+#include <kj/compat/brotli.h>
+#include <kj/compat/gzip.h>
+#include <kj/one-of.h>
+
+namespace workerd::api::streams {
+
+namespace {
+// Used to consume and collect all data from a ReadableStreamSource up to a specified
+// limit. Throws if the limit is exceeded before EOF.
+class AllReader final {
+ public:
+  explicit AllReader(ReadableStreamSource& input, size_t limit): input(input), limit(limit) {
+    JSG_REQUIRE(limit > 0, TypeError, "Memory limit exceeded before EOF.");
+    KJ_IF_SOME(length, input.tryGetLength(rpc::StreamEncoding::IDENTITY)) {
+      // Oh hey, we might be able to bail early.
+      JSG_REQUIRE(length <= limit, TypeError, "Memory limit would be exceeded before EOF.");
+    }
+  }
+  KJ_DISALLOW_COPY_AND_MOVE(AllReader);
+
+  kj::Promise<kj::Array<const kj::byte>> readAllBytes() {
+    co_return co_await read<kj::byte>();
+  }
+
+  kj::Promise<kj::String> readAllText() {
+    co_return kj::String(co_await read<char>(ReadOption::NULL_TERMINATE));
+  }
+
+ private:
+  ReadableStreamSource& input;
+  size_t limit;
+
+  enum class ReadOption {
+    NONE,
+    NULL_TERMINATE,
+  };
+
+  template <typename T>
+  kj::Promise<kj::Array<T>> read(ReadOption option = ReadOption::NONE) {
+    // Read in chunks and accumulate them. Use an exponential growth strategy
+    // to determine chunk sizes to minimize the number of iterations and
+    // allocations on large streams.
+    kj::Vector<kj::Array<T>> parts;
+    size_t runningTotal = 0;
+    // TODO(later): Make these configurable someday?
+    static constexpr size_t MIN_BUFFER_CHUNK = 1024;
+    static constexpr size_t DEFAULT_BUFFER_CHUNK = 4096;
+    static constexpr size_t MAX_BUFFER_CHUNK = DEFAULT_BUFFER_CHUNK * 4;
+
+    // If we know in advance how much data we'll be reading, then we can attempt to optimize the
```

Comment:
@harrishancock ... re-implemented with an exponential growth strategy as suggested. Should work a whole lot better.

Reply:
Nice! And it looks slightly simpler, too.

We might want to reconsider the size of `MAX_BUFFER_CHUNK` in the future. Consider reading a 1 MiB stream -- it'll have 68 allocations, I think, whereas if we lifted the max chunk size to 1 MiB or more, it'd be 11.

But, that's just configuration that we can bikeshed.